### PR TITLE
Migrate website

### PR DESCRIPTION
**Pull request by anandubajith**
Edit: 
- Everything has been migrated

--- 
- The original URL from which the content is taken is not displayed on top ( if available in page metadata )
- Docker based setup of the Documentation site was added
- Few pages have been migrated
  - FAQ - https://anandubajith.github.io/expos-v2/faq/
  - About - https://anandubajith.github.io/expos-v2/about/
  - Documentation - https://anandubajith.github.io/expos-v2/documentation/
- Starter html for pages to migrate have been added

### CODE REVIEW

File: docs/about.md [hunk 1]
```diff
@@ -1 +1,139 @@
...
+
+It must be emphasized that the project by no means a replacement to a theory course in operating systems,
+but is only designed to suppliment it.
+
+eXpOS evolved from an earlier version (called <a href="http://xosnitc.github.io/" target="_blank">XOS</a>) which was successfully used for undergraduate instruction at the department of Computer Science and Engineering, NIT Calicut. eXpOS extends XOS with multi­user support, co­operative multi­tasking, interrupt based disk transfer and interprocess communication facility. Both these platforms were developed by teams of undergraduate CSE students of NIT Calicut through a series of undergraduate major projects. <!--Maintenance and user support for the system are being provided by a team of volunteers through a <a href="https://groups.google.com/forum/#!forum/xos-users" target="_blank">user's mailing list</a>-->
+
+### Authors
```
Comment (ishangh) [inline docs/about.md:78]: Move Authors License and Acknowledgement to different pages if required

File: docs/about.md [hunk 2]
```diff
@@ -1 +1,139 @@
...
+    5.  support for multiple users
+    6.  demand-paged memory management
+3.  The OS code has to be completely written by the student with no built-in code given, except for a bootstrap ROM code and code implementing an [application level library interface](abi.html#library). The programming support has been kept minimal to avoid high level programming tools hiding too many low level issues from the OS programmer. Essentially, the student has to do “low level programming” on a “high level hardware".
+4.  The system is designed to help the student appreciate the functioning of a “sufficiently complex” operating system and **not** to help him/her gain experience with real operating systems programming. The architecture, the OS, the API, ABI etc. are all different from any of the existing real systems or standards in use, but designed to make the task of programming the OS as simple as possible.
+
+---
+
+<p>Since the operating system kernel is a software that forms a layer between the hardware and the application program, the first major hurdle that a student faces while writing an OS is in understanding the complexities of the underlying hardware.</p>
```
Comment (ishangh) [inline docs/about.md]: p tags remove them if they are not required

File: docs/arch-spec/instruction_set.html [hunk 3]
```diff
@@ -0,0 +1,640 @@
+<!--=========== BEGIN SECTION ================-->
+<section id="instruction_set">
```
Comment (ishangh) [inline docs/arch-spec/instruction_set.html]: Will these remain as HTML?
Replace filenames to hyphens. https://developers.google.com/style/filenames

File: docs/arch-spec/virtual_machine_spec.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+<p>The virtual (or logical) memory addresses that can be generated by a user mode program is determined by the value
+of the PTLR register.  The virtual address space of a user mode program is a contiguous address space starting from
+0 to 512*PTLR-1. The virtual addreses generated by a user mode program are translated into physical addresses by the <a href="../arch_spec-files/paging_hardware.html" target="_blank">address translation scheme</a> of XSM. The user mode program is oblivious to the address translation.</p>
+
+<p>Operating systems using the XSM architecture typically specifies the maximum value of PTLR permitted
+for its user processes and divide the allocated address space for an application into library, code,
+data, stack, heap etc.</p>
+
+<figure>
```
Comment (ishangh) [inline docs/arch-spec/virtual_machine_spec.md]: Images can be centered using a markdown format too. See index.md

File: docs/assets/css/extra.css [hunk 5]
```diff
@@ -0,0 +1,13 @@
+.stage-link {
```
Comment (ishangh) [inline docs/assets/css/extra.css:1]: The colors do not contrast in dark mode for links

**Comment by anandubajith** [comment]
Roadmap Stages 1 - 12 ( Preparatory stages ) have been migrated

https://anandubajith.github.io/expos-v2/roadmap/

**Comment by anandubajith** [comment]
Most of the files in `os_design/*`has expandable annotations inside code blocks
eg: http://exposnitc.github.io/os_design-files/exec.html

How should they be handled? Custom CSS?

I also found this https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations

UPDATE:
This is how it'll look with custom CSS overriding the default styles for `details` and `summary` tags inside code blocks
https://anandubajith.github.io/expos-v2/modules/module_07/

shall i proceed with this?

File: docs/about.md [hunk 6]
```diff
@@ -1 +1,138 @@
...
+
+- [Separation of concerns](https://en.wikipedia.org/wiki/Edsger_W._Dijkstra)
+- [Things shall be made as simple as possible, but no simpler](https://en.wikipedia.org/wiki/Albert_Einstein)
+- ["Build to learn" rather than "learn to build"](https://en.wikipedia.org/wiki/Fred_Brooks)
+
+These thoughts have translated into the following decisions:
+
+1.  Keep the upward and downward interfaces of the OS kernel idealistically simple – This means:
+    1.  The architecture platform on which the OS is implemented is kept simple to understand and program, yet capable of supporting virtual memory, interrupt driven disk/IO, time sharing, exception handling, demand paging etc.
+    2.  The application programming interface ([API](http://exposnitc.github.io/os_spec.html#collapse1)) (and the application binary interface [­(ABI)](http://exposnitc.github.io/abi.html) ) provided by the OS are kept simple and minimal.
```
Comment (ishangh) [inline docs/about.md]: Update links

File: docs/about.md [hunk 7]
```diff
@@ -1 +1,138 @@
...
+
+### Authors
+The content in the website and the documentation has been authored in the Department of Computer Science and Engineering, <a href="http://nitc.ac.in" target="_blank">National Institute of Technology, Calicut</a> under the guidance of Dr. Murali Krishnan K. The project's activity started in the year 2010 and is currently under progress. Below is a list of co-authors and contributors to the project. The work evolved from an earlier version of the project called the <a href="https://xosnitc.github.io" target="_blank">XOS project</a> under the guidance of Dr. Murali Krishnan K.
```
Comment (ishangh) [inline docs/about.md]: a tags can be replaced with `[text](link){target=_blank}`

File: docs/arch-spec/index.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
+---
+title: XSM Architecture Specification
+original_url: https://exposnitc.github.io/arch_spec.html
+hide:
```
Comment (ishangh) [inline docs/arch-spec/index.md:4]: I think we should keep the webpages consistent. Would it be a problem if all pages except the homepage have both navigation and toc? even if there is one element in toc.

File: docs/arch-spec/index.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+The basic data storage components of the machine are **registers**, **memory** and the **disk**. The memory and the disk are organised as sequences of words. The word is the fundamental memory/disk storage element and each memory/disk word can store a string. In the following, we often (mis)use notation and use the term “word” to refer to the “string stored in the memory/disk word”.
+
+!!! note "Important Note"
+    _Througout the architecture documentation, we have used the terms kernel mode for previleged mode and user mode for unprevileged mode of machine execution. However, "kernel" and "user" are OS level abstractions and not connected with the hardware. Hence the above usage is incorrect technically. However since the kernel of an OS normally runs in the previliged mode and user programs execute in the unpreviliged mode, the (incorrect) terminology has been used._
+
+This document contains the following sections:
+
+#### [Machine Organization](machine_organisation.md)
```
Comment (ishangh) [inline docs/arch-spec/index.md]: We can replace filenames to hyphens. https://developers.google.com/style/filenames

File: docs/arch-spec/instruction_set.md [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+Every instruction in XSM is 2 words long. The instructions provided by the XSM architecture can be classified into **unprivileged (user)** and **privileged (kernel)** instructions. The semantics of each instruction is different, depending on the mode in which the machine is running.
+
+
+## Unprivileged Instructions
+
+XSM provides a set of unprivileged instructions, which are the only instructions available when the machine is executing in the user mode. The machine can execute unprivileged instructions in kernel mode also, however, the instruction semantics will be different as explained in [Paging hardware](paging_hardware.html). The unprivileged instructions are _Data Transfer Instructions_, _Arithmetic Instructions_, _Logical Instructions_, _Stack Instructions_, _Sub-routine instructions_, _Debug instructions_ and _Software interrupts_. Registers available in user mode are R0-R19, SP, BP and IP.
```
Comment (ishangh) [inline docs/arch-spec/instruction_set.md]: Update links to markdown

File: docs/arch-spec/instruction_set.md [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+
+  
+  
+
+For all the above instructions, Ri/Rj may be any register **except IP**.
+
+!!! note
+	Only registers R0-R19, SP and BP shall be used in code that executes in unpreviliged mode. An [exception](interrupts_exception_handling.html#exceptions) (illegal instruction) will be generated otherwise.
+
+### [Logical Instructions]
```
Comment (ishangh) [inline docs/arch-spec/instruction_set.md]: Requirement for the square bracket?

File: docs/arch-spec/instruction_set.md [hunk 12]
```diff
@@ -0,0 +1,274 @@
+---
```
Comment (ishangh) [inline docs/arch-spec/instruction_set.md]: Add nesting in elements. It's difficult to distinguish between instructions and their description. Descriptions can be on the next line after semantics. No need to add an extra space as the space makes it look like a separate instruction.

File: docs/arch-spec/instruction_set.md [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+### [HALT]
+
+_Syntax_ : HALT
+
+_Semantics_ : This instruction causes the machine to halt immediately.
+
+  
+  
+
+##  <span style="color:red">NEXSM Additional Privileged Instructions</span>
```
Comment (ishangh) [inline docs/arch-spec/instruction_set.md]: Heading levels 2 and 3 are not distinguishable. Try with 2 and 4

File: docs/arch-spec/interrupts_exception_handling.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+Note: Internally XSM machine has an _interrupt vector table_ that maps each interrupt number to a corresponding address (of the interrupt handler). The vector table is stored starting from physical address 492 of the ROM memory. Locations 492,493,494 and 495 stores the addresses of exception handler, timer interrupt handler, disk interrupt handler and console interrupt handler. Location 496 contains the address of the first software interrupt handler - INT 4, location 497 contains address of INT 5 handler and so on. When the machine encounters an INT n instruction, the corresponding ROM location is searched for the handler address and IP is set to this value. Since the values are hard-coded in the ROM code of the XSM simulator given to you, you cannot change the addresses of these handlers. See [here](../arch_spec-files/machine_organisation.html#Boot ROM) for more details about Boot ROM and interrupt vector table.
+
+!!! note
+    XSM does not allow the INT instruction to be executed in kernel mode.
+
+
+### **Timer Interrupt**
```
Comment (ishangh) [inline docs/arch-spec/interrupts_exception_handling.md]: You used size 2 heading in instruction_set.md why does it start with size 3 here?

File: docs/arch-spec/interrupts_exception_handling.md [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+The XSM architecture includes a timer that can cause periodic interrupts during execution in unprivileged mode. The timer ticks with each unprivileged instruction executed and once the number of ticks reaches a threshold specified during boot-time (see --timer option of the [XSM simulator](../support_tools-files/xsm-simulator.html)), the tick counter is reset and a timer interrupt is raised. At that point, the machine executes the following actions:
+
+1.  :red_circle: Increment the value of the SP register and push the value in the IP register into the memory address pointed to by SP. Since the machine is executing in unpreviliged mode, the address in SP is treated as a logical address and the machine performs [address translation](paging_hardware.html) to calculate the physical address corresponding to the SP value.
```
Comment (ishangh) [inline docs/arch-spec/interrupts_exception_handling.md]: update links

File: docs/feedback/nitc18/index.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
+---
+hide:
+    - toc
```
Comment (ishangh) [inline docs/feedback/nitc18/index.md:3]: Change html

**Comment by ishangh** [comment]
Try to migrate everything first. ignore the cleanup comments we can do those later. I haven't looked through everything, will do so later.

File: docs/arch-spec/instruction_set.html (hunk 3 above)
Comment (anandubajith) [reply docs/arch-spec/instruction_set.html]: - All files have been renamed for consistency

File: docs/feedback/nitc18/index.md (hunk 16 above)
Comment (anandubajith) [reply docs/feedback/nitc18/index.md:3]: - feedback pages have been migrated to markdown

**Comment by anandubajith** [comment]
All pages have been migrated

TODO:

- [x] Replace internal links, image links 
- [ ] Prune dead pages ( links which are not referenced anywhere )  (eg: Old module pages are still there, having them show up in search might confuse user )

File: docs/assets/css/extra.css (hunk 5 above)
Comment (anandubajith) [reply docs/assets/css/extra.css:1]: light mode
dark mode

**Comment by anandubajith** [comment]
All links have been rewritten

note:
- temporarily hiding the edit button on all pages so that broken link checkers can be used

**Comment by anandubajith** [comment]
Everything has been migrated

**Comment by ishangh** [comment]
LGTM bugs acn be fixed if raised.

**Pull request merged**